server/service/src/sync/translation_and_integration.rs

```diff
@@ -0,0 +1,206 @@
+                    self.sync_buffer
+                        .record_integration_error(&sync_record, &error)?;
+                    result.insert_error(
+                        &sync_record.table_name,
+                        Some(&sync_record),
+                        Some(&integration_records),
+                        error,
+                    );
+                }
+            }
+        }
+        Ok(result)
+    }
+}
+
+impl IntegrationRecords {
+    pub(crate) fn integrate(&self, connection: &StorageConnection) -> Result<(), RepositoryError> {
+        for upsert in self.upserts.iter() {
+            upsert.upsert(connection)?
+        }
+
+        Ok(())
+    }
+}
+
+impl PullUpsertRecord {
+    pub(crate) fn upsert(&self, con: &StorageConnection) -> Result<(), RepositoryError> {
+        use PullUpsertRecord::*;
+        match self {
+            Name(record) => NameRowRepository::new(con).upsert_one(record),
+            Unit(record) => UnitRowRepository::new(con).upsert_one(record),
+            Item(record) => ItemRowRepository::new(con).upsert_one(record),
+            Store(record) => StoreRowRepository::new(con).upsert_one(record),
+            MasterList(record) => MasterListRowRepository::new(con).upsert_one(record),
+            MasterListLine(record) => MasterListLineRowRepository::new(con).upsert_one(record),
+            MasterListNameJoin(record) => MasterListNameJoinRepository::new(con).upsert_one(record),
+            Report(record) => ReportRowRepository::new(con).upsert_one(record),
+            Number(record) => NumberRowRepository::new(con).upsert_one(record),
+            Location(record) => LocationRowRepository::new(con).upsert_one(record),
+            StockLine(record) => StockLineRowRepository::new(con).upsert_one(record),
+            NameStoreJoin(record) => NameStoreJoinRepository::new(con).upsert_one(record),
+            Invoice(record) => InvoiceRowRepository::new(con).upsert_one(record),
+            InvoiceLine(record) => InvoiceLineRowRepository::new(con).upsert_one(record),
+            Stocktake(record) => StocktakeRowRepository::new(con).upsert_one(record),
+            StocktakeLine(record) => StocktakeLineRowRepository::new(con).upsert_one(record),
+            Requisition(record) => RequisitionRowRepository::new(con).upsert_one(record),
+            RequisitionLine(record) => RequisitionLineRowRepository::new(con).upsert_one(record),
+        }
+    }
+}
+
+impl TranslationAndIntegrationResults {
+    fn new() -> TranslationAndIntegrationResults {
+        Default::default()
+    }
+
+    fn insert_error(
+        &mut self,
+        table_name: &str,
+        sync_record: Option<&SyncBufferRow>,
+        integration_records: Option<&IntegrationRecords>,
+        error: anyhow::Error,
+    ) {
+        let entry = self
+            .0
+            .entry(table_name.to_owned())
+            .or_insert(Default::default());
+        entry.errors_count += 1;
+        entry.errors.push((
```

Comment:
Is errors_count needed? or is there a case where errors_count += errors.len()?

Reply:
Thanks but I've actually removed the errors from TranslationAndIntegrationResults, it wasn't actually useful (i just wanted a summary at the end of each sync to show error and integration count against each record, and the full error was deleted.